### DISCUSSION TITLE
MicroOS version freezing <-> Kube-Hetzner tags

### DISCUSSION THREAD

**Discussion by valkenburg-prevue-ch** [mysticaltech/terraform-hcloud-kube-hetzner#634] [Ideas]
As per this issue https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/issues/623 and its fix, https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/commit/74d1a60487bbf893eea21a600a4a3161b6ef71ad , I think we should try to find a way to make explicit which version of MicroOS is tied to which release of kube-hetzner. Right now this is undocumented, and the only relatively certain way is to "check out all the latest today", i.e. the latest kube-hetzner release and the latest microos image.

That is not really safe, right?

I do now know what the way out it is, but I'm hoping to open a discussion about this.

**Comment by valkenburg-prevue-ch** [comment]
@mysticaltech @aleksasiriski Let's take this here. Replying to https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/issues/623#issuecomment-1463707442 ,

it's actually the second time in three months.

> https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/issues/555
[EDIT: should have been https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/issues/528 ]
https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/issues/623

**Reply by mysticaltech** [reply]
@valkenburg-prevue-ch No #555 was not a microOS issue, it was due to an external package that we install!

**Reply by valkenburg-prevue-ch** [reply]
Oops, so sorry, I copied the wrong link, don't know why, but I meant:
https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/issues/528

**Reply by mysticaltech** [reply]
About #623, if it's not clear, before it did not detect the extra network interface automatically, so it was a missing feature, now they fixed it, our code base was just not robust enough.

MicroOS is designed NOT TO BREAK. If you want to know more, great presentation by the creator where he explains everything you need to know: https://www.youtube.com/watch?v=idZEJ0OYfWU

**Reply by valkenburg-prevue-ch** [reply]
I get why microos is awesome, and I'm not suggesting we trade it in, I would just like to find a solution that adds stability. Like simply listing the microos-image against which a certain release is guaranteed to work. I'm asking a question, not proposing a solution.

**Reply by valkenburg-prevue-ch** [reply]
Please see my OP opening this discussion.

**Reply by mysticaltech** [reply]
I understand, no worries. There's no version to my knowledge it's a rolling release, so we always get the latest and greatest. Normally, we should be quite stable from now on from the OS side of things, as we do not use anything special, just the basics.

**Reply by valkenburg-prevue-ch** [reply]
But the fact that it is unversioned goes against the standards of fully testing updates before you deploy them. Does the creator address that in the video? (Will watch later).

**Reply by mysticaltech** [reply]
But that's the point, it's designed to remove that requirement for stability. (If I remember well, he makes this fact pretty clear in his explanation).

**Reply by valkenburg-prevue-ch** [reply]
Well, he goes a great length to explain why updates should be atomic and offer the possibility to rollback. Which is how microos works once installed, and I love that part. But the way we use it, downloading the latest image in realtime, is the opposite of that. Nothing is atomic, nothing can be rolled back. Installing the exact same code base today and tomorrow may have a very different outcome. That makes me and I can imagine many users a bit scared. For me, I'm good, I host a fixed image on a private server, so I can in fact roll back. But the kube-hetzner project publicly cannot.

**Reply by mysticaltech** [reply]
I know that @ifeulner does that too. It's fine and that's what the URL param is for, people can choose whatever mirror they want, even their own. Personally, I like to stick with the latest and greatest, and I am sure enough that it won't break (and would be curious if it did break to make sure to make it more robust, but it almost never happens). 

That said, with the upcoming release, we will work via an initial snapshot, so unless you delete and recreate the snapshot, you will have the assurance you seek for the users, for the life of the cluster! 🚀🚀

**Comment by mysticaltech** [comment]
Upcoming LeapMicro will solve those worries.
